pkg/admission/pod_eviction_test.go

```diff
@@ -0,0 +1,576 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	admissionv1 "k8s.io/api/admission/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/grafana/rollout-operator/pkg/config"
+	"github.com/grafana/rollout-operator/pkg/zpdb"
+)
+
+const (
+	statefulSetZoneA  = "ingester-zone-a"
+	statefulSetZoneB  = "ingester-zone-b"
+	statefulSetZoneC  = "ingester-zone-c"
+	testPodZoneA0     = "ingester-zone-a-0"
+	testPodZoneA1     = "ingester-zone-a-1"
+	testPodZoneA2     = "ingester-zone-a-2"
+	testPodZoneB0     = "ingester-zone-b-0"
+	testPodZoneB1     = "ingester-zone-b-1"
+	testPodZoneB2     = "ingester-zone-b-2"
+	testPodZoneC0     = "ingester-zone-c-0"
+	testPodZoneC1     = "ingester-zone-c-1"
+	testPodZoneC2     = "ingester-zone-c-2"
+	testNamespace     = "test-dev-0"
+	rolloutGroupValue = "ingester"
+	rolloutGroupLabel = config.RolloutGroupLabelKey
+	nameLabel         = "name"
+)
+
+type testContext struct {
+	ctx      context.Context
+	logs     *dummyLogger
+	client   *fake.Clientset
+	request  admissionv1.AdmissionReview
+	pdbCache *zpdb.ZpdbCache
+	podCache *zpdb.PodEvictionCache
+}
+
+// A dummyLogger accumulates log lines into a slice so we can assert that certain logs were recorded
+type dummyLogger struct {
+	logs []string
+}
+
+func newDummyLogger() *dummyLogger {
+	return &dummyLogger{}
+}
+
+func (d *dummyLogger) Log(keyVals ...interface{}) error {
+	var parts []string
+	for i := 0; i < len(keyVals); i += 2 {
+		key := fmt.Sprint(keyVals[i])
+		var val string
+		if i+1 < len(keyVals) {
+			val = fmt.Sprint(keyVals[i+1])
+		} else {
+			val = "<missing>"
+		}
+		parts = append(parts, fmt.Sprintf("%s=%s", key, val))
+	}
+
+	d.logs = append(d.logs, strings.Join(parts, " "))
+	return nil
+}
+
+// assertHasLog ensures there is a log line which contains all the given elements
+// Note that we do not care about the ordering of the elements
+func (d *dummyLogger) assertHasLog(t *testing.T, elements []string) {
+
+	for _, line := range d.logs {
+		fmt.Printf("%s\n", line)
```

Comment:
Is this still needed? If you'd like the tests to emit logs, then perhaps `dummyLogger.Log` would be a better place for this.